components/nonmem/src/output_files/shk.rs

```diff
@@ -54,210 +74,321 @@ impl ShkTable {
         }
 
         lines.join("\n")
     }
 }
 
-#[derive(Debug, Default, Clone)]
-pub struct ShkReader;
+// Converts RawShkTable into vec of ShkTable based
+// on subpopulation.
+impl From<RawShkTable> for Vec<ShkTable> {
+    fn from(raw: RawShkTable) -> Self {
+        // Find all unique subpopulation numbers
+        let mut subpops: Vec<usize> = raw.rows.iter().map(|row| row.subpop).collect();
+        subpops.sort_unstable();
+        subpops.dedup();
+
+        // Create one ShkTable per subpopulation
+        let mut tables: Vec<ShkTable> = subpops
+            .iter()
+            .map(|&subpop| {
+                let mut table = ShkTable::new(subpop);
+                table.method = raw.method;
+                table
+            })
+            .collect();
+
+        // Process each raw row and populate the appropriate semantic fields
+        for row in &raw.rows {
+            if let Some(table) = tables.iter_mut().find(|t| t.subpop == row.subpop) {
+                table.set_field_by_type_num(row.type_num, row.values.clone());
+            }
+        }
+
+        tables
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ShkReader {
+    only_last: bool,
+    only_method: Option<EstimationMethod>,
+}
+
+impl Default for ShkReader {
+    fn default() -> Self {
+        ShkReader {
+            only_last: false,
+            only_method: None,
+        }
+    }
+}
 
 impl ShkReader {
-    pub fn parse_file(&self, path: impl AsRef<Path>) -> Result<Vec<Vec<ShkTable>>> {
+    /// Only keep the table corresponding to the given method.
+    /// Exclusive of `only_last`: it will be set to false if this method is called
+    pub fn only_method(mut self, method: EstimationMethod) -> Self {
+        self.only_method = Some(method);
+        self.only_last = false;
+        self
+    }
+
+    /// Keep all tables from the input file.
+    /// This will disable `only_last` and `only_method`
+    pub fn keep_all_tables(mut self) -> Self {
+        self.only_last = false;
+        self.only_method = None;
+        self
+    }
+
+    /// Only keep the last table.
+    /// This is the default.
+    pub fn only_last(mut self) -> Self {
+        self.only_last = true;
+        self.only_method = None;
+        self
+    }
+
+    /// Parse a .shk file and return semantic table structure
+    pub fn parse_file_semantic(&self, path: impl AsRef<Path>) -> Result<Vec<Vec<ShkTable>>> {
+        let raw_tables = self.parse_file(path)?;
+        Ok(raw_tables.into_iter().map(|raw| raw.into()).collect())
+    }
+
+    pub fn parse_file(&self, path: impl AsRef<Path>) -> Result<Vec<RawShkTable>> {
         let path = path.as_ref();
         let file = fs::File::open(path)?;
         let reader = BufReader::new(file);
         let tables = self.parse(reader)?;
         Ok(tables)
     }
 
-    pub fn parse<R: BufRead>(&self, reader: R) -> Result<Vec<Vec<ShkTable>>> {
+    pub fn parse<R: BufRead>(&self, mut reader: R) -> Result<Vec<RawShkTable>> {
```

Comment:
`parse` is probably the wrong name if we are not parsing anything anymore. `read`?